linera-core/src/chain_worker/state.rs

```diff
@@ -405,23 +398,26 @@ where
                 WorkerError::UnneededValue { value_hash }
             );
         }
-        let tasks = self
+        let locations = self
             .recent_hashed_certificate_values
             .subtract_cached_items_from::<_, Vec<_>>(
                 required_locations_left.into_values(),
                 |location| &location.certificate_hash,
             )
             .await
             .into_iter()
-            .map(|location| {
-                self.storage
-                    .contains_hashed_certificate_value(location.certificate_hash)
-                    .map(move |result| (location, result))
-            })
             .collect::<Vec<_>>();
+        let hashes = locations
+            .iter()
+            .map(|location| location.certificate_hash)
+            .collect::<Vec<_>>();
+        let results = self
+            .storage
+            .contains_hashed_certificate_values(hashes)
+            .await?;
         let mut missing_locations = vec![];
-        for (location, result) in future::join_all(tasks).await {
-            if !result? {
+        for (location, result) in locations.into_iter().zip(results) {
```

Comment:
(I see: Here we aren't `zip`ping with the keys themselves. I guess the function could take an `Fn` that maps the caller-level to the storage-level keys… not sure if that would be better.)

Reply:
I am not sure what is the issue here. Is there an issue with zipping?